### PR TITLE
Dunitz/908 create genesets table

### PR DESCRIPTION
**Pull request by MDunitz**
#### Reviewers
**Functional:** 
- @Bento007 

**Readability:** 
 - @seve 

---

## Changes
- add
    - Geneset table and corresponding entity
    - GenesetDatasetLink table and corresponding entity
    - DB migration file
- modify
   - Dataset table to include genesets field (m2m to geneset through GenesetDatasetLink)
   - Collections table to include genesets field (Backref)

## Definition of Done (from ticket)
- Can add a geneset to the database
- Can retrieve a geneset from the database
- Can retrieve a list of all genesets linked to a collection
- Can add a geneset_dataset_link
- Can retrieve a geneset_dataset_link
- Can retrieve all genesets for a dataset

### CODE REVIEW

File: backend/database/README.md [hunk 1]
```diff
@@ -56,10 +56,18 @@ export DEPLOYMENT_STAGE=test
...
+cd backend/
 make db/import FROM=${}
 ```
-where from is the name of the *.sqlc* file downloaded. You may need to run this a few times, until there are no significant errors.
-1. Run the migration test:
+where from is the name of the *.sqlc* file downloaded. For example 
+```shell script
+make db/import FROM=corpora_dev-202102221309
+```
+- Note: The file is stored under backend/database/file_name but the make command will look in import/file_name this is due to the way [the local paths are mapped to the docker container](3.console.aws.amazon.com/s3/buckets/hosted-cellxgene-dev/?region=us-west-2&tab=overview)
```
Comment (Bento007) [inline backend/database/README.md:66]: Your link is broken.

**Review by Bento007 — APPROVED** [review]
LGTM, fix your link

File: backend/database/README.md (hunk 1 above)
Comment (MDunitz) [reply backend/database/README.md]: updated! thanks

File: tests/unit/backend/utils.py [hunk 2]
```diff
@@ -1,3 +1,6 @@
+import random
```
Comment (Bento007) [inline tests/unit/backend/utils.py:1]: If using random make sure you seed it with the current time.

File: tests/unit/backend/utils.py [hunk 3]
```diff
@@ -53,3 +56,25 @@ def get(cls, **kwargs):
...
+            gene_symbols=gene_symbols,
+            collection_id="test_collection_id",
+            collection_visibility=CollectionVisibility.PUBLIC.name,
+        )
+
+        bogus_data.update(**kwargs)
+        return bogus_data
+
+    @staticmethod
+    def generate_random_string(self, length=7):
```
Comment (Bento007) [inline tests/unit/backend/utils.py:79]: remove self for static methods

**Review by Bento007 — CHANGES_REQUESTED** [review]
Since `geneset.name` is not required yet it must be unique, can you have multiple genesets named Null in a collection?

File: backend/corpora/common/corpora_orm.py [hunk 4]
```diff
@@ -361,3 +364,37 @@ class DbDatasetProcessingStatus(Base, AuditMixin):
...
+    collection = relationship("DbCollection", uselist=False, back_populates="genesets")
+    datasets = relationship("DBGenesetDatasetLink", back_populates="geneset", cascade="all, delete-orphan")
+
+    __table_args__ = (
+        ForeignKeyConstraint([collection_id, collection_visibility], [DbCollection.id, DbCollection.visibility]),
+        UniqueConstraint("name", "collection_id", "collection_visibility", name="_geneset_name__collection_uc"),
+    )
+
+
+class DBGenesetDatasetLink(Base, AuditMixin):
```
Comment (Bento007) [inline backend/corpora/common/corpora_orm.py:390]: change to DbGenesetDatasetLink

**Review by seve — APPROVED** [review]
LGTM

**Review by Bento007 — APPROVED** [review]
LGTM

**Pull request merged**